File: python/aurigraph_sdk/namespaces/contracts.py
"""Namespace for /api/v11/active-contracts and /api/v11/contract-bindings endpoints."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any
from urllib.parse import quote

if TYPE_CHECKING:
    from aurigraph_sdk.client import AurigraphClient


class ContractsApi:
    """Smart contract and Ricardian contract operations."""

    def __init__(self, client: AurigraphClient) -> None:
        self._client = client

    def deploy(self, contract: dict[str, Any]) -> dict[str, Any]:
        """Deploy a new contract.

        Args:
            contract: The contract deployment payload.

        Returns:
            Deployment receipt.
        """
        return self._client._post("/contracts", contract)

    def invoke(self, contract_id: str, method: str, args: dict[str, Any] | None = None) -> dict[str, Any]:
        """Invoke a method on a deployed contract.

        Args:
            contract_id: The contract to invoke.
            method: The method name to call.
            args: Optional method arguments.

        Returns:
            Invocation result.
        """
        enc = quote(contract_id, safe="")
        body: dict[str, Any] = {"method": method}
        if args:
            body["args"] = args
        return self._client._post(f"/contracts/{enc}/invoke", body)

    def get(self, contract_id: str) -> dict[str, Any]:
        """Get contract details by ID.

        Args:
            contract_id: The contract identifier.

        Returns:
            Contract details.
        """
        enc = quote(contract_id, safe="")
        return self._client._get(f"/active-contracts/{enc}")

    def get_tokens(self, contract_id: str) -> list[dict[str, Any]]:
        """Get tokens bound to a contract.

        Args:
            contract_id: The contract identifier.

        Returns:
            List of token bindings.
        """
        enc = quote(contract_id, safe="")
        data = self._client._get(f"/active-contracts/{enc}/tokens")
        if isinstance(data, list):  # type: ignore[arg-type]
            return data  # type: ignore[return-value]
        if isinstance(data, dict):
            return data.get("tokens", data.get("items", data.get("data", [])))
        return []

    def bind_composite(self, request: dict[str, Any]) -> dict[str, Any]:
        """Bind a composite token to a contract.

        Args:
            request: The composite bind request payload.

        Returns:
            Binding result.
        """
        return self._client._post("/contract-bindings/composite", request)

    def get_bindings_for_contract(self, contract_id: str) -> list[dict[str, Any]]:
        """Get all bindings for a contract.

        Args:
            contract_id: The contract identifier.

        Returns:
            List of composite bindings.
        """
        enc = quote(contract_id, safe="")
        data = self._client._get(f"/contract-bindings/contract/{enc}")
        if isinstance(data, list):  # type: ignore[arg-type]
            return data  # type: ignore[return-value]
        if isinstance(data, dict):
            return data.get("bindings", data.get("items", data.get("data", [])))
        return []

File: rust/src/namespaces/contracts.rs
//! Smart contract and Ricardian contract namespace for `/api/v11/contracts/*` endpoints.
//!
//! Provides contract deployment, invocation, and retrieval.
//!
//! # Example
//!
//! ```rust,no_run
//! # async fn example(client: &aurigraph_sdk::AurigraphClient) -> aurigraph_sdk::Result<()> {
//! let deployed = client.contracts().deploy("RicardianLease", &serde_json::json!({"tenant": "Alice"})).await?;
//! println!("Deployed: {}", deployed);
//!
//! let contract = client.contracts().get("contract-1").await?;
//! println!("Contract: {}", contract);
//! # Ok(())
//! # }
//! ```

use crate::client::AurigraphClient;
use crate::error::Result;

/// Contracts API namespace.
///
/// Obtained via [`AurigraphClient::contracts()`].
#[derive(Debug)]
pub struct ContractsApi<'a> {
    client: &'a AurigraphClient,
}

impl<'a> ContractsApi<'a> {
    pub(crate) fn new(client: &'a AurigraphClient) -> Self {
        Self { client }
    }

    /// Deploy a new smart contract or Ricardian contract.
    ///
    /// Calls `POST /api/v11/contracts/deploy`.
    pub async fn deploy(
        &self,
        template: &str,
        params: &serde_json::Value,
    ) -> Result<serde_json::Value> {
        let body = serde_json::json!({
            "template": template,
            "parameters": params,
        });
        self.client.post("/contracts/deploy", &body).await
    }

    /// Invoke a function on a deployed contract.
    ///
    /// Calls `POST /api/v11/contracts/{contract_id}/invoke`.
    pub async fn invoke(
        &self,
        contract_id: &str,
        function: &str,
        args: &serde_json::Value,
    ) -> Result<serde_json::Value> {
        let body = serde_json::json!({
            "function": function,
            "arguments": args,
        });
        self.client
            .post(&format!("/contracts/{}/invoke", contract_id), &body)
            .await
    }

    /// Get a contract by ID.
    ///
    /// Calls `GET /api/v11/contracts/{contract_id}`.
    pub async fn get(&self, contract_id: &str) -> Result<serde_json::Value> {
        self.client
            .get(&format!("/contracts/{}", contract_id))
            .await
    }
}
